### ISSUE TITLE
[enc2] Newly created user has no valid private key

### ISSUE THREAD

**Issue by PVince81** [owncloud/core#15981]
### Steps
1. Enable encryption in the admin page
2. Enable the default encryption module
3. Log out then log in again as admin
4. Create a new user "user1"
5. Log in as "user1"

### Expected result
User1 can start using their account

### Actual result
Warning "Encryption App enabled but your keys are not initialized, please log-out and log-in again"
The "welcome.txt" file from the skeleton as 0 bytes.

### Logs
(this was when creating another user "user2")
```
{"reqId":"6trBRL4Y9aIDpXbGNe4T","remoteAddr":"127.0.0.1","app":"files_skeleton","message":"copying skeleton for user2 from \/srv\/www\/htdocs\/owncloud\/core\/skeleton to \/user2\/files\/","level":0,"time":"2015-04-30T15:11:32+00:00","method":"POST","url":"\/owncloud\/"}
{"reqId":"6trBRL4Y9aIDpXbGNe4T","remoteAddr":"127.0.0.1","app":"hook","message":"error while running hook (OCA\\Encryption\\Hooks\\UserHooks::login): Private Key missing for user: please try to log-out and log-in again","level":3,"time":"2015-04-30T15:11:33+00:00","method":"POST","url":"\/owncloud\/"}
```

### Versions
ownCloud 8.1 git master (41d7870d7b8aa1b635fd1d2d589322160bb46d31)

@schiesbn @th3fallen @DeepDiver1975

**Comment by PVince81** [comment]
Introduced by https://github.com/owncloud/core/pull/15968 (@icewind1991)

It looks like changing the hook from "setup" to "preSetup" messes up with the creation.

I wonder if it messes up with other things too, risky change :frowning:

**Comment by schiesbn** [comment]
I debugged this a bit. The problem seems that now when we write the welcome.txt on the first login the encryption wrapper is already enabled and therefore we try to encrypt the welcome.txt. But this fails because the encryption keys are not yet initialized which can can only do after the filesystem was loaded because we need to read the key from the users data directory.

I don't really know how to fix this at the moment. But I have the feeling that the quick fix here https://github.com/owncloud/core/pull/15968/files (switch from the setup hook to the pre-setup hook) was wrong. The right fix for https://github.com/owncloud/core/issues/15640 would probably be to make sure that we always sort the wrappers according to their priorities correctly and also call the sort algorithm everytime a new wrapper gets added, like described here https://github.com/owncloud/core/issues/15640#issuecomment-95504123

cc @icewind1991 can you please look into this issue once more? Thanks!

**Comment by icewind1991** [comment]
Trying to keep the wrappers sorted when adding wrappers to already wrapped storages requires to much complexity imo.

An easier way would be to bypass the wrappers when copying the skeleton or delay copying the skeleton until the filesystem (and all it's wrappers) are ready.

**Comment by DeepDiver1975** [comment]
> delay copying the skeleton until the filesystem (and all it's wrappers) are ready.

sounds the most reasonable approach - @icewind1991 can you please come up with a PR on this? THX

**Comment by schiesbn** [comment]
hm, I see the performance argument. On the other hand I don't feel comfortable to know that even if proper priorities are added we still can't guarantee that the wrappers are in the right order.

**Comment by PVince81** [comment]
Beware that doing any deferred FS operations need to be done only **ONCE**, which means we need to make sure that concurrent parallel requests that might trigger the lazy code only happens once, maybe through some kind of locking, maybe a flag in the DB that tells whether the skeleton was copied. And set that flag before starting the actual copy.

**Comment by PVince81** [comment]
This PR fixes it: https://github.com/owncloud/core/pull/16075
